MetaMorpheus/EngineLayer/ClassicSearch/ClassicSearchEngine.cs

```diff
@@ -227,18 +227,23 @@ private void AddPeptideCandidateToPsm(ScanWithIndexAndNotchInfo scan, double thi
                 // valid hit (met the cutoff score); lock the scan to prevent other threads from accessing it
                 lock (Locks[scan.ScanIndex])
                 {
-                    bool scoreImprovement = PeptideSpectralMatches[scan.ScanIndex] == null || (thisScore - PeptideSpectralMatches[scan.ScanIndex].RunnerUpScore) > -SpectralMatch.ToleranceForScoreDifferentiation;
+                    bool scoreImprovement = SpectralMatches[scan.ScanIndex] == null || (thisScore - SpectralMatches[scan.ScanIndex].RunnerUpScore) > -SpectralMatch.ToleranceForScoreDifferentiation;
 
-                    if (scoreImprovement)
+                    if (!scoreImprovement) 
+                        return;
+
+                    // if the PSM is null, create a new one; otherwise, add or replace the peptide
+                    switch (SpectralMatches[scan.ScanIndex])
                     {
-                        if (PeptideSpectralMatches[scan.ScanIndex] == null)
-                        {
-                            PeptideSpectralMatches[scan.ScanIndex] = new PeptideSpectralMatch(peptide, scan.Notch, thisScore, scan.ScanIndex, ArrayOfSortedMS2Scans[scan.ScanIndex], CommonParameters, matchedIons);
-                        }
-                        else
-                        {
-                            PeptideSpectralMatches[scan.ScanIndex].AddOrReplace(peptide, thisScore, scan.Notch, CommonParameters.ReportAllAmbiguity, matchedIons);
-                        }
+                        case null when GlobalVariables.AnalyteType == AnalyteType.Oligo:
+                            SpectralMatches[scan.ScanIndex] = new OligoSpectralMatch(peptide, scan.Notch, thisScore, scan.ScanIndex, ArrayOfSortedMS2Scans[scan.ScanIndex], CommonParameters, matchedIons);
+                            break;
+                        case null:
+                            SpectralMatches[scan.ScanIndex] = new PeptideSpectralMatch(peptide, scan.Notch, thisScore, scan.ScanIndex, ArrayOfSortedMS2Scans[scan.ScanIndex], CommonParameters, matchedIons);
+                            break;
+                        default:
+                            SpectralMatches[scan.ScanIndex].AddOrReplace(peptide, thisScore, scan.Notch, CommonParameters.ReportAllAmbiguity, matchedIons);
+                            break;
                     }
```

Comment:
This doesn't seem like it should be a switch statement, as you're not really switching on anything.

`if (PeptideSpectralMatches[scan.ScanIndex] == null)`
`   PeptideSpectralMatches[scan.ScanIndex] = GlobalVariables.AnalyteType == AnalyteType.Oligo ? new OSM : new PSM`
`else`
`     SpectralMatches[scan.ScanIndex].AddOrReplace`

Would make it easier to understand whats happening here